src/audio/dp_queue.c

```diff
@@ -0,0 +1,367 @@
+	if (dp_queue_is_shared(dp_queue))
+		dp_queue_invalidate_shared(dp_queue,
+					   (__sparse_force void __sparse_cache *)*data_ptr,
+					   req_size);
+
+	return 0;
+}
+
+static int dp_queue_release_data(struct sof_source *source, size_t free_size)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_source(source);
+
+	if (free_size) {
+		/* data consumed, free buffer space, no need for any special cache operations */
+		k_spinlock_key_t key = dp_queue_lock(dp_queue);
+
+		dp_queue->available_data -= free_size;
+		dp_queue->free_space += free_size;
+		dp_queue->read_offset =
+			(dp_queue->read_offset + free_size) % dp_queue->data_buffer_size;
+
+		dp_queue_unlock(dp_queue, key);
+	}
+
+	return 0;
+}
+
+static int dp_queue_set_ipc_params(struct dp_queue *dp_queue,
+				   struct sof_ipc_stream_params *params,
+				   bool force_update)
+{
+	if (dp_queue->hw_params_configured && !force_update)
+		return 0;
+
+	dp_queue->audio_stream_params.frame_fmt = params->frame_fmt;
+	dp_queue->audio_stream_params.rate = params->rate;
+	dp_queue->audio_stream_params.channels = params->channels;
+	dp_queue->audio_stream_params.buffer_fmt = params->buffer_fmt;
+
+	dp_queue->hw_params_configured = true;
+
+	return 0;
+}
+
+/*
+ * note! a sparse warning will be generated here till
+ * https://github.com/thesofproject/sof/pull/8013 is implemented
+ */
+
+static int dp_queue_set_ipc_params_source(struct sof_source *source,
+					  struct sof_ipc_stream_params *params,
+					  bool force_update)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_source(source);
+
+	return dp_queue_set_ipc_params(dp_queue, params, force_update);
+}
+
+static int dp_queue_set_ipc_params_sink(struct sof_sink *sink,
+					struct sof_ipc_stream_params *params,
+					bool force_update)
+{
+	struct dp_queue *dp_queue = dp_queue_get_from_sink(sink);
+
+	return dp_queue_set_ipc_params(dp_queue, params, force_update);
+}
+
+
+/*
+ * note! a sparse warning will be generated here till
+ * https://github.com/thesofproject/sof/pull/8013 is implemented
+ */
+static const struct source_ops dp_queue_source_ops = {
+	.get_data_available = dp_queue_get_data_available,
+	.get_data = dp_queue_get_data,
+	.release_data = dp_queue_release_data,
+	.audio_set_ipc_params = dp_queue_set_ipc_params_source,
+};
+
+static const struct sink_ops dp_queue_sink_ops = {
+	.get_free_size = dp_queue_get_free_size,
+	.get_buffer = dp_queue_get_buffer,
+	.commit_buffer = dp_queue_commit_buffer,
+	.audio_set_ipc_params = dp_queue_set_ipc_params_sink,
+};
+
+struct dp_queue *dp_queue_create(uint32_t ibs, uint32_t obs, uint32_t flags,
+				 struct sof_audio_stream_params *external_stream_params)
+{
+	uint32_t align;
+	struct dp_queue *dp_queue;
+
+	/* allocate DP structure */
+	if (flags & DP_QUEUE_MODE_SHARED)
+		dp_queue = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+				   sizeof(*dp_queue));
+	else
+		dp_queue = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*dp_queue));
```

Comment:
Just logging a comment that this is not compatible with spinlock use, so needs to be modified. You pointed this yourself already, but just to keep track.